### DISCUSSION TITLE
Problem during Gap8 bootloader

### DISCUSSION THREAD

**Discussion by ClementSrs** [bitcraze/discussions#1024] [Q&A]
Hello,
I'm starting to work on Crazyflie for my internship on Ubuntu 20.04. I realised that crazyflie had an old bootloader because when I flash an example from the tutorial, it freezes at 99%. 
However, when I try to build and flash the bootloader using STLink-v2 this error appears and I don't know how to solve it... 

$ git clone https://github.com/bitcraze/aideck-gap8-bootloader.git
$ cd aideck-gap8-bootloader
$ docker run --rm -it -v $PWD:/module/ --device /dev/ttyUSB0 --privileged -P bitcraze/aideck /bin/bash -c 'export GAPY_OPENOCD_CABLE=interface/ftdi/olimex-arm-usb-tiny-h.cfg; source /gap_sdk/configs/ai_deck.sh; cd /module/; make all image flash'

---
rm -f /module/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/flash.img
    CC main.c
    CC com.c
    CC cpx.c
src/cpx.c: In function 'cpxPrintToConsole':
src/cpx.c:67:19: warning: pointer targets in passing argument 1 of 'vsnprintf_' differ in signedness [-Wpointer-sign]
   len = vsnprintf(consoleTx.data, sizeof(consoleTx.data), fmt, ap);
                   ^~~~~~~~~
In file included from /gap_sdk/rtos/freeRTOS/demos/gwt/config/gap8/FreeRTOS_util.h:46:0,
                 from /gap_sdk/rtos/freeRTOS/vendors/gwt/pmsis/include/pmsis_types.h:35,
                 from /gap_sdk/rtos/freeRTOS/vendors/gwt/pmsis/include/pmsis.h:53,
                 from src/cpx.c:30:
/gap_sdk/rtos/freeRTOS/vendors/gwt/libs/printf/printf.h:87:5: note: expected 'char *' but argument is of type 'uint8_t * {aka unsigned char *}'
 int vsnprintf_(char* buffer, size_t count, const char* format, va_list va);
     ^~~~~~~~~~
    CC bl.c
    CC flash.c
    CC FreeRTOS_util.c
    ASM  gap8_iet.S
    ASM  startup_gap8.S
    ASM  cluster_core.S
    ASM  asm_util.S
    ASM  port_asm.S
    CC  gap8_it.c
    CC  system_gap8.c
    CC  gap_io.c
    CC  errno.c
    CC  stdlib.c
    CC  string.c
    CC  cl_malloc.c
    CC  cl_to_fc_delegate.c
    CC  fc_to_cl_delegate.c
    CC  cl_team.c
    CC  hyperbus_cl_internal.c
    CC  uart_cl_internal.c
    CC  cl_dma_irq.c
    CC  fc_event.c
    CC  fll.c
    CC  gpio.c
    CC  pad.c
    CC  pmu.c
    CC  pmu_internal.c
    CC  pwm.c
    CC  pwm_internal.c
    CC  rtc.c
    CC  rtc_internal.c
    CC  timer.c
    CC  hyperbus.c
    CC  hyperbus_internal.c
    CC  cpi.c
    CC  cpi_internal.c
    CC  dmacpy.c
    CC  dmacpy_internal.c
    CC  i2c.c
    CC  i2c_internal.c
    CC  i2s.c
    CC  i2s_internal.c
    CC  spi.c
    CC  spi_internal.c
    CC  uart.c
    CC  uart_internal.c
    CC  flash.c
    CC  partition.c
    CC  flash_partition.c
    CC  md5.c
    CC  fs.c
    CC  lfs.c
    CC  lfs_util.c
    CC  pi_lfs.c
    CC  read_fs.c
    CC  semihost.c
    CC  host_fs.c
    CC  ota.c
    CC  ota_utility.c
    CC  updater.c
    CC  bootloader_utility.c
    CC  ai_deck.c
    CC  camera.c
    CC  himax.c
    CC  hyperflash.c
    CC  hyperram.c
    CC  spiram.c
    CC  spiflash.c
/tmp/ccUfuXYW.s: Assembler messages:
/tmp/ccUfuXYW.s:120: Warning: ignoring changed section attributes for .data
    CC  ram.c
    CC  alloc_extern.c
    CC  pi_log.c
    CC  event_kernel.c
    CC  cl_l1_malloc.c
    CC  fc_l1_malloc.c
    CC  l2_malloc.c
    CC  malloc_external.c
    CC  malloc_internal.c
    CC  pi_malloc.c
    CC  mem_slab.c
    CC  device.c
    CC  pmsis_task.c
    CC  pmsis_backend_native_task_api.c
    CC  port.c
    CC  printf.c
    CC  list.c
    CC  queue.c
    CC  tasks.c
    CC  timers.c
    CC  event_groups.c
    CC  stream_buffer.c
gapy --target=ai_deck --platform=board --work-dir=/module/BUILD/GAP8_V2/GCC_RISCV_FREERTOS    run --image --binary=/module/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/bootloader 
gapy --target=ai_deck --platform=board --work-dir=/module/BUILD/GAP8_V2/GCC_RISCV_FREERTOS    run --flash --binary=/module/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/bootloader 
gapy --target=ai_deck  gen_flash_image --boot-loader /module/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/bootloader  -o /module/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/flash.img 
make: Nothing to be done for 'image'.
gapy --target=ai_deck --platform=board --work-dir=/module/BUILD/GAP8_V2/GCC_RISCV_FREERTOS    run --flash --force --binary=/module/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/bootloader 
Flashing image with command:
gap8-openocd -d0 -c "gdb_port disabled; telnet_port disabled; tcl_port disabled" -c "script /module/stlink-v2.cfg; script target/gap8revb.tcl; script tcl/flash_image.tcl; script tcl/jtag_boot.tcl; gap_flash_raw_hyper /module/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/target.board.devices.flash.img 75808 /gap_sdk/utils/openocd_tools; exit;"
Open On-Chip Debugger 0.10.0+dev-gb84d97ec4 (2022-04-20-14:26)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
debug_level: 0

DEPRECATED! use 'adapter driver' not 'interface'
DEPRECATED! use 'adapter speed' not 'adapter_khz'
DEPRECATED! use 'adapter driver' not 'interface'
TARGET create
GAP8 INIT TARGET
Error: no device found
Error: unable to open ftdi device with vid 0483, pid 3748, description 'ST-LINK/V2', serial '*' at bus location '*'

**Traceback (most recent call last):
  File "/gap_sdk/utils/gapy/gapy", line 162, in <module>
    main()
  File "/gap_sdk/utils/gapy/gapy", line 152, in main
    operationFunc(args, config, system)
  File "/gap_sdk/utils/gapy/run.py", line 184, in operationFunc
    raise RuntimeError('Runner has failed with value: %d' % status)
RuntimeError: Runner has failed with value: -1
make: *** [/gap_sdk/rtos/freeRTOS/vendors/gwt/rules/freeRTOS_rules.mk:404: flash] Error 1**
---

It doesn't seem to detect the STLink-v2 even though it appears in my computer's USB ports.

$Bus 001 Device 048: ID 0483:3748 STMicroelectronics ST-LINK/V2

There is the stlink-v2.cfg I use : 
**---
interface ftdi
ftdi_device_desc "ST-LINK/V2"
ftdi_vid_pid 0x0483 0x3748

ftdi_layout_init 0x0808 0x0a1b
ftdi_layout_signal nSRST -oe 0x0200
ftdi_layout_signal nTRST -data 0x0100 -oe 0x0100
ftdi_layout_signal LED -data 0x0800
---**

Also I added the debugger to the udev rules : 
SUBSYSTEM=="usb", ATTRS{idVendor}=="0483", ATTRS{idProduct}=="3748", MODE="0664", GROUP="plugdev".

Another thing that may be impacting the error is that when I update the crazyflie firmware, the console shows ESP32: I (505) SYS: Initialized instead of ESP32: I (910) SYS: Initialized.

Thanks in advance

**Comment by Janamejay-Sharma** [comment]
Hi,

The issue appears to be `--device /dev/ttyUSB0`. For Ubuntu 20.04, that isn't the correct format.

- Connect your JTAG device to your computer via USB.
- Run lsusb. Find the JTAG device in the list and take note of its **Bus** & **Device** number. For example, it may read `Bus 001 Device 010`, so the required path would be `/dev/bus/usb/001/010`.

Hope this helps!
Regards,
Jay

**Comment by boomer319** [comment]
This is your last command line input:
$ docker run --rm -it -v $PWD:/module/ --device /dev/ttyUSB0 --privileged -P bitcraze/aideck /bin/bash -c 'export GAPY_OPENOCD_CABLE=interface/ftdi/**olimex-arm-usb-tiny-h.cfg**; source /gap_sdk/configs/ai_deck.sh; cd /module/; make all image flash'

You were executing the command for the Olimex. [This](https://github.com/analogdevicesinc/openocd/tree/master/tcl/interface) is the collection of supported devices.


I asked a question regarding this too: https://github.com/orgs/bitcraze/discussions/942#discussioncomment-7194235

Try this out:

$ docker run --rm -it -v $PWD:/module/ --device /dev/ttyUSB0 --privileged -P bitcraze/aideck /bin/bash -c 'export GAPY_OPENOCD_CABLE=interface/**stlink-v2.cfg**; source /gap_sdk/configs/ai_deck.sh; cd /module/; make all image flash'
